shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -411,29 +410,38 @@ - (void)loadAOTData:(NSString*)assetsDir {
 }
 
 - (void)setViewController:(FlutterViewController*)controller {
-  if (_viewController != controller) {
+  if (_viewController == controller) {
+    // From nil to nil, or from non-nil to the same controller.
+    return;
+  }
+  if (_viewController == nil && controller != nil) {
+    NSAssert(controller.engine == nil,
+             @"Failed to set view controller to the engine: "
+             @"The given FlutterViewController is already attached to an engine %@. "
+             @"If you wanted to create an FlutterViewController and set it to an existing engine, "
+             @"you should create it with init(engine:, nibName, bundle:) instead.",
+             controller.engine);
     _viewController = controller;
-
-    if (_semanticsEnabled && _bridge) {
-      _bridge->UpdateDefaultViewController(_viewController);
-    }
-
-    if (!controller && !_allowHeadlessExecution) {
+    [_viewController attachToEngine:self withId:kFlutterDefaultViewId];
+  } else if (_viewController != nil && controller == nil) {
```

Comment:
~Similarly, we know that _viewController is non-nil due to the check right above this.~

Similar comment as above; I think I'd leave this check as-is but add the comment.
```suggestion
  } else if (_viewController != nil && controller == nil) {
    // From non-nil to nil.
```